### PR TITLE
[ENG-6619] validate icon PNGs before running build

### PR DESCRIPTION
**Pull request by dsokal**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

From https://linear.app/expo/issue/ENG-6619/validate-pngs-before-running-build:

> "INVALID_PNG_SIGNATURE" error seems like it may be caused by the user providing an invalid png file for their app icon or splash screen. For managed apps, we should be able to validate this before running their build, making this much less frustrating and we can provide a nice error at that point in the process. 

# How

For Android, validate the following app config fields:
- exp.icon
- exp.android.icon
- exp.android.adaptiveIcon.foregroundImage
- exp.android.adaptiveIcon.backgroundImage
- exp.splash.image
- exp.notification.icon

For iOS:
- exp.icon
- exp.ios.icon
- exp.splash.image
- exp.notification.icon

plus check whether `exp.ios.icon ?? ctx.exp.icon` is not tranparent.

# Test Plan

Tests.

### CODE REVIEW

File: packages/eas-cli/src/build/validate.ts [hunk 1]
```diff
@@ -50,3 +58,61 @@ export async function checkGoogleServicesFileAsync<T extends Platform>(
...
+    if (err instanceof ImageNonPngError) {
+      Log.error(`The iOS app icon must be a PNG file.`);
+      Log.error(`${chalk.bold(icon)} is not a PNG file`);
+      Errors.exit(1);
+    } else if (err instanceof ImageTransparencyError) {
+      Log.error(
+        `Your app icon can't have transparency if you wish to upload your app to the Apple App Store.`
+      );
+      Log.error(`Read more here: https://expo.fyi/remove-alpha-channel`);
+      Errors.exit(1);
```
Comment (dsokal) [inline packages/eas-cli/src/build/validate.ts:113]: @brentvatne can you please review the error messages?

File: packages/eas-cli/src/build/validate.ts [hunk 2]
```diff
@@ -50,3 +58,61 @@ export async function checkGoogleServicesFileAsync<T extends Platform>(
...
+  if (!ctx.exp.android?.adaptiveIcon?.foregroundImage) {
+    return;
+  }
+
+  const { foregroundImage } = ctx.exp.android.adaptiveIcon;
+  if (!foregroundImage.endsWith('.png')) {
+    Log.error(`The Android Adaptive Icon foreground image must be a PNG file.`);
+    Log.error(`expo.android.adaptiveIcon.foregroundImage = ${chalk.bold(foregroundImage)}`);
+    Errors.exit(1);
+  }
```
Comment (szdziedzic) [inline packages/eas-cli/src/build/validate.ts:86]: We are quickly checking the Android icon using `endsWith` method. Do we want to do the same for the iOS icon?

File: packages/eas-cli/src/build/validate.ts [hunk 3]
```diff
@@ -50,3 +57,68 @@ export async function checkGoogleServicesFileAsync<T extends Platform>(
...
+      Log.error(
+        `Your iOS app icon (${iconConfigPath}) can't have transparency if you wish to upload your app to the Apple App Store.`
+      );
+      Log.error(learnMore('https://expo.fyi/remove-alpha-channel', { dim: false }));
+      Errors.exit(1);
+    } else {
+      throw err;
+    }
+  }
+}
```
Comment (dsokal) [inline packages/eas-cli/src/build/validate.ts:124]: @brentvatne can you please review the error messages?

File: packages/eas-cli/src/build/validate.ts [hunk 4]
```diff
@@ -50,3 +57,68 @@ export async function checkGoogleServicesFileAsync<T extends Platform>(
...
+async function validateAndroidIconAsync(ctx: CommonContext<Platform.ANDROID>): Promise<void> {
+  if (!ctx.exp.android?.adaptiveIcon?.foregroundImage) {
+    return;
+  }
+
+  const { foregroundImage } = ctx.exp.android.adaptiveIcon;
+  if (!foregroundImage.endsWith('.png')) {
+    Log.error(`The Android Adaptive Icon foreground image must be a PNG file.`);
+    Log.error(`"expo.android.adaptiveIcon.foregroundImage" = ${foregroundImage}`);
+    Errors.exit(1);
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/build/validate.ts:84]: discussed offline, my mistake `Errors.exit` throws and error not process.exit,

**Review by wkozyra95 — APPROVED** [review]

**Review by szdziedzic — APPROVED** [review]

**Comment by EvanBacon** [comment]
This seems like a bug in Prebuild. The icon generation step should be converting the inputs into compliant outputs. Also icon fields can be URLs and a variety of other image formats which will now suddenly not work with EAS Build after this change lands.

**Comment by brentvatne** [comment]
- ✅ the fields being validated for Android look correct to me and this should help a lot to prevent the ~30-40 failed builds per day that we see
- ⚠️ I believe it's not necessary to validate on iOS - I tried a couple builds and prebuild appears to be resilient against different file types. That said, if we don't have these validations in EAS CLI and an image is invalid, it would be annoying to spin up a job and get all the way to where the build fails in prebuild (or later) to give feedback.

Some follow up here for @byCedric / @EvanBacon on other CLI tooling:
- `expo doctor` warns about non-PNG image files for icon fields, even those these are actually supported on iOS (even if we don't recommend them). We need to decide if we want to explicitly support non-PNG files and unify behavior.
- Android icons should also be made similarly resilient in prebuild if we decide that we want to uniformly support non-PNG files, [I made a task for this](https://linear.app/expo/issue/ENG-6829/make-prebuild-for-android-resilient-against-different-image-asset)

**Comment by dsokal** [comment]
@EvanBacon 

> Also icon fields can be URLs and a variety of other image formats which will now suddenly not work with EAS Build after this change lands.

Can you elaborate? My code supports URLs. See https://github.com/expo/eas-cli/pull/1477/files#diff-e1c7866672b8c773c57f0c858398754872eebbf02a3ee3e8b238f45396efb312R72

> This seems like a bug in Prebuild.

I don't know anything about Prebuild. This PR is based on the Linear task description. Happy to see this fixed in Prebuild.
Also, fixing this in prebuild doesn't prevent users from running redundant builds that will fail because of incorrect images.

@brentvatne 

> let's comment out iOS validation for now - I think it's possible that we'll revisit it in the future. otherwise looks good to ship

Sounds good!

**Review by wkozyra95 — APPROVED** [review]

**Review by szdziedzic — APPROVED** [review]

**Comment by brentvatne** [comment]
@dsokal 

> What about transparency?

We handle removing transparency automatically in prebuild: https://github.com/expo/expo/blob/15d2337a623a7921f68d4e800cf10cf38c95f051/packages/%40expo/prebuild-config/src/plugins/icons/withIosIcons.ts#L130-L137

I tested this and it worked as expected. Where we may still encounter problems here is if people have bare apps where they set the icon manually and it has transparency

**Comment by brentvatne** [comment]
@dsokal - one other suggestion - can we scope these validations to only projects that are using sdk 46 and lower? @EvanBacon has improved prebuild so it should be able to handle non-png icons and convert them to png.

**Comment by EvanBacon** [comment]
The feature looks good, but adding a gate to only apply on managed projects less than SDK 47 would be good.

**Comment by dsokal** [comment]
@brentvatne @EvanBacon can you please review the PR again?

**Review by wkozyra95 — APPROVED** [review]

**Review by brentvatne — APPROVED** [review]

**Pull request merged**